### PR TITLE
Make ViewComponent::Collection a collection

### PR DESCRIPTION
**Pull request by sammyhenningsson**
### Summary

Despite the name `ViewComponent::Collection`, is a thing that renders a collection, rather than being a collection itself. This is a bit confusing and prevents some use cases were we need to operate on the collection in more ways than just rendering them.

I think my use case (see background below) is quite unique. However I do still believe that it could be useful to be able to do more with collections, than to render them 😄  

### Background

My current use case is that I have users that has some set of data and they should be able to select how this data is presented. This will be done through selecting a template implemented using a view component that gets instantiated with the users data. In order to be sure that the view component (template) can render I've added validations on them. Something like:
```ruby
# app/components/application_component.rb                                                                                                   
class ApplicationComponent < ViewComponent::Base                                                                                            
  include ActiveModel::Validations                                                                                                          
end   
                                                                                                                                            
# app/components/site_component.rb
class SiteComponent < ApplicationComponent
  attr_reader :site                                                            
  validates :site, :name, presence: true                                                                                                    
  validate :valid_sections
        
  def initialize(site:)                                                                                                                     
    @site = site                                                                                                                            
  end 
        
  def name                                                                                                                                  
    site.name                                                                                                                               
  end 
        
  def sections                                                                                                                              
    SectionComponent.with_collection(site.sections)                                                                                         
  end 
                                                                                                                                            
  private 
        
  def valid_sections
    sections.each do |section|                                                                                                              
      errors.add(:sections, :invalid) unless section.valid?                                                                                 
    end                                                                                                                                     
  end                                                                                                                                       
end

# app/components/section_component.rb
class SectionComponent < ApplicationComponent                                                                                               
  validates :section, :foobar, presence: true                                                                                               
                                                                                                                                            
  def initialize(section:)
    @section = section
  end
  
  def foobar                                                                                                                                
    site.foobar
  end
end
```

The reason that I need to validate the data at runtime, is because I will add more and more templates that my users can select. But they need to know which ones they can use and what data they need to change to make use of the ones that are currently invalid.

### CODE REVIEW

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,9 @@ title: Changelog
 # Changelog
 
 ## main
+* Make `ViewComponent::Collection` act like a collection of view components
```
Comment (Spone) [inline docs/CHANGELOG.md:9]:
```suggestion

* Make `ViewComponent::Collection` act like a collection of view components
```

File: docs/index.md [hunk 2]
```diff
@@ -178,6 +178,7 @@ ViewComponent is built by over a hundred members of the community, including:
 <img src="https://avatars.githubusercontent.com/wdrexler?s=64" alt="wdrexler" width="32" />
 <img src="https://avatars.githubusercontent.com/mattwr18?s=64" alt="mattwr18" width="32" />
 <img src="https://avatars.githubusercontent.com/dylanatsmith?s=64" alt="dylanatsmith" width="32" />
+<img src="https://avatars.githubusercontent.com/sammyhenningsson?s=64" alt="dylanatsmith" width="32" />
```
Comment (Spone) [inline docs/index.md:181]:
```suggestion
<img src="https://avatars.githubusercontent.com/sammyhenningsson?s=64" alt="sammyhenningsson" width="32" />
```

File: docs/CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,10 @@ title: Changelog
 
 ## main
 
+* Make `ViewComponent::Collection` act like a collection of view components
```
Comment (joelhawksley) [inline docs/CHANGELOG.md:10]:
```suggestion
* Make `ViewComponent::Collection` act like a collection of view components.
```

**Review by joelhawksley — APPROVED** [review]

**Comment by joelhawksley** [comment]
@sammyhenningsson looks good! I'll merge once the conflicts are resolved ❤️

**Pull request merged**